services/director-v2/src/simcore_service_director_v2/cli/core.py

```diff
@@ -0,0 +1,298 @@
+    settings: AppSettings, node_id: NodeIDStr
+) -> AnyHttpUrl:
+    dynamic_sidecar_names = DynamicSidecarNamesHelper.make(UUID(node_id))
+    hostname = dynamic_sidecar_names.service_name_dynamic_sidecar
+    port = settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR.DYNAMIC_SIDECAR_PORT
+    return parse_obj_as(AnyHttpUrl, f"http://{hostname}:{port}")  # NOSONAR
+
+
+async def _save_node_state(
+    app,
+    dynamic_sidecar_client: api_client.DynamicSidecarClient,
+    retry_save: int,
+    node_uuid: NodeIDStr,
+    label: str = "",
+) -> None:
+    typer.echo(f"Saving state for {node_uuid} {label}")
+    async for attempt in AsyncRetrying(
+        wait=wait_random_exponential(),
+        stop=stop_after_attempt(retry_save),
+        reraise=True,
+    ):
+        with attempt:
+            typer.echo(f"Attempting to save {node_uuid} {label}")
+            await dynamic_sidecar_client.save_service_state(
+                _get_dynamic_sidecar_endpoint(app.state.settings, node_uuid)
+            )
+
+
+async def async_project_save_state(project_id: ProjectID, retry_save: int) -> None:
+    async with _initialized_app() as app:
+        projects_repository: ProjectsRepository = get_repository(
+            app, ProjectsRepository
+        )
+        project_at_db = await projects_repository.get_project(project_id)
+
+        typer.echo(f"Saving project '{project_at_db.uuid}' - '{project_at_db.name}'")
+
+        dynamic_sidecar_client = api_client.get_dynamic_sidecar_client(app)
+        nodes_failed_to_save: list[NodeIDStr] = []
+        for node_uuid, node_content in project_at_db.workbench.items():
+            # onl dynamic-sidecars are used
+            if not await requires_dynamic_sidecar(
+                service_key=node_content.key,
+                service_version=node_content.version,
+                director_v0_client=DirectorV0Client.instance(app),
+            ):
+                continue
+
+            try:
+                await _save_node_state(
+                    app,
+                    dynamic_sidecar_client,
+                    retry_save,
+                    node_uuid,
+                    node_content.label,
+                )
+            except Exception:  # pylint: disable=broad-except
+                nodes_failed_to_save.append(node_uuid)
+
+    if nodes_failed_to_save:
+        typer.echo(
+            "The following nodes failed to save:"
+            + "\n- "
+            + "\n- ".join(nodes_failed_to_save)
+            + "\nPlease try to save them individually!"
+        )
+        sys.exit(1)
+
+    typer.echo(f"Save complete for project {project_id}")
+
+
+### NODE SAVE STATE
+
+
+async def async_node_save_state(node_id: NodeID, retry_save: int) -> None:
+    async with _initialized_app() as app:
+        dynamic_sidecar_client = api_client.get_dynamic_sidecar_client(app)
+        await _save_node_state(
+            app, dynamic_sidecar_client, retry_save, NodeIDStr(f"{node_id}")
```

Comment:
Q: no way to get the node label beforehand?

Reply:
No not in this case, I do not want you to provide the project_id just to fetch a label here. It is not worth the effort.